scheduler/src/cook/hooks.clj

```diff
@@ -0,0 +1,174 @@
+  (reify SchedulerHooks
+    (check-job-submission-default [_] default-accept)
+    (check-job-submission [_ _] default-accept)
+    (check-job-invocation [_ _] default-accept)))
+
+(defn resolve-symbol
+  "Resolve the given symbol to the corresponding Var."
+  [sym]
+  (resolve (some-> sym namespace symbol use) sym))
+
+(defn create-default-hook-object
+  "Returns the hook object. If no hook factory defined, returns an always-accept hook object."
+  [config]
+  (let [{:keys [settings]} config
+        {:keys [hook-factory]} settings
+        {:keys [factory-fn arguments]} hook-factory]
+    (log/info (str "Setting up hooks with factory config: " hook-factory " and factory-fn " factory-fn))
+    (if factory-fn
+      (do
+        (if-let [resolved-fn (resolve-symbol (symbol factory-fn))]
+          (do
+            (log/info (str "Resolved as " resolved-fn " with " arguments))
+            (resolved-fn arguments))
+          (throw (ex-info "Unable to resolve factory function" (assoc hook-factory :ns (namespace factory-fn))))))
+      accept-all-hook)))
+
+;  Contains the hook object that matches to a given job map. This code may create a new hook object or re-use an existing one.
+(mount/defstate hook-object
+  :start (create-default-hook-object config))
+
+(mount/defstate submission-hook-batch-timeout-seconds
+  :start (-> config :settings :hooks :submission-hook-batch-timeout-seconds t/seconds))
+(mount/defstate age-out-last-seen-deadline-minutes
+  :start (-> config :settings :hooks :age-out-last-seen-deadline-minutes t/minutes))
+(mount/defstate age-out-first-seen-deadline-minutes
+  :start (-> config :settings :hooks :age-out-first-seen-deadline-minutes t/minutes))
+(mount/defstate age-out-seen-count
+  :start (-> config :settings :hooks :age-out-seen-count))
+
+; We may see up to the entire scheduler queue, so have a big cache here.
+; This is called in the scheduler loop. If it hasn't been looked at in more than 2 hours, the job has almost assuredly long since run.
+(def ^Cache job-invocations-cache
+  (-> (CacheBuilder/newBuilder)
+      (.maximumSize 100000)
+      (.expireAfterAccess 2 TimeUnit/HOURS)
+      (.build)))
+
+
+(defn aged-out?
+  "Looking at the cached dictionary, Has this job been in the queue long enough that we should force it to run and get it done. This includes:
+  1. Being in the queue for a long time (hours)
+  2. Attempting to launch the job fairly recently (seen in last several minutess)
+  3. Tried multiple times to get the job to launch."
+  [{:keys [last-seen first-seen seen-count] :as old-result}]
+  {:post [(or (true? %) (false? %))]}
+  (let [last-seen-deadline (->> age-out-last-seen-deadline-minutes
+                                (t/minus- (t/now)))
+        first-seen-deadline (->> age-out-first-seen-deadline-minutes
+                                 (t/minus- (t/now)))]
+    (and
+      (boolean old-result)
```

Comment:
Is this effectively `(not (nil? old-result))`? If so, can we use that instead?

Reply:
Done.